packages/react/src/components/MultiSelect/MultiSelect.tsx

```diff
@@ -393,6 +400,23 @@ const MultiSelect = React.forwardRef(
       });
     }, [items]);
 
+    const itemsWithSelectAll = selectAll
+      ? [selectAllOption, ...filteredItems]
+      : filteredItems;
```

Comment:
Hi Taylor, Please find the response:

_Does the user get the selectAllOption in the items returned through onChange?_
Yes user will get selectAllOption in the items returned through onChange.

_Should they get that back?_
It's the part of Items now.

_How would someone using the controlled component check the select all, or mark it indeterminate? Do they need to or does it happen automatically?_
If they are using controlled component, they won't have access to selectAll or indeterminate state. To enable these things user has to enable 'hasSelectAll' as true.

_If there are unavoidable prop conflicts, like selectAll just plain doesn't work with the component in a controlled state, then ideally we'd flag this with a warning in the dev console._
If we are adding the prop 'hasSelectAll' prop in controlled state, it will work simply as the default one. Only issue is we can have conflict of logics, for eg - the current controlled component is selecting disabled items as well while we are not. 

While writing these responses, I found that we have to disable select All functionality in case of controlled state. User can either use selectedItems or hasSelectAll prop. We will add a console warning as suggested.